src/main/kotlin/com/example/springrestuser/controllers/PrivateAssetController.kt

```
@@ -0,0 +1,92 @@
+import com.example.springrestuser.dto.UserDTO
+import com.example.springrestuser.enums.AssetSubtype
+import com.example.springrestuser.enums.AssetType
+import com.example.springrestuser.services.PrivateAssetService
+import com.example.springrestuser.services.UserAssetService
+import org.hibernate.exception.DataException
+import org.springframework.web.bind.annotation.*
+import java.lang.Exception
+import java.util.*
+
+@RestController
+class PrivateAssetController(
+    private val service: PrivateAssetService,
+) {
+
+    @PostMapping("/asset/private/")
+    fun add(@RequestBody asset: PrivateAssetDTO): PrivateAssetDTO {
+        if (validate(asset)) {
+            return toDTO(service.add(toDO(asset)))
+        } else {
+            throw RuntimeException("Invalid input data")
+        }
+    }
+
+    @GetMapping("/asset/private/{id}")
+    fun get(@PathVariable id: Int): PrivateAssetDTO {
+        return toDTO(service.get(id))
+    }
+
+    @GetMapping("/asset/private/")
+    fun getAll(): List<PrivateAssetDTO> {
+        val list = service.getAll()
+        val res = mutableListOf<PrivateAssetDTO>()
+        for (asset in list) {
+            res.add(toDTO(asset))
+        }
+        return res
+    }
+
+    @PutMapping("/asset/private/{id}")
+    fun modify(@PathVariable id: Int, @RequestBody asset: PrivateAssetDTO) {
+        asset.id = id
+        if (validate(asset)) {
+            service.modify(toDO(asset))
+        } else {
+            throw RuntimeException("Invalid input data")
+        }
+
+    }
+
+    @DeleteMapping("/asset/private/{assetId}")
+    fun remove(@PathVariable assetId: Int) {
+        service.delete(assetId)
+    }
+
+    fun validate(asset: PrivateAssetDTO): Boolean {
+        var res = true
+        if (asset.price < 0) {
+            res = false
+        }
+        if (asset.type.lowercase(Locale.getDefault()) != "private") {
+            res = false
+        }
+        return res
+    }
```

you could throw an exception here so that you won't need to introduce business logic in the controller